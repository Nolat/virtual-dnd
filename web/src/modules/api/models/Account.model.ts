import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn
} from "typeorm";

import User from "./User.model";

@Entity()
export default class Account extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: string;

  @OneToOne(() => User)
  @JoinColumn()
  user: User;

  @Column()
  providerType: string;

  @Column()
  providerId: string;

  @Column()
  providerAccountId: string;

  @Column({ nullable: true })
  refreshToken: string;

  @Column({ nullable: true })
  accessToken: string;

  @Column({ type: "timestamp", nullable: true })
  accessTokenExpiresOn: Date;

  @CreateDateColumn({ type: "timestamp" })
  createdAt: Date;

  @UpdateDateColumn({ type: "timestamp" })
  updatedAt: Date;
}
