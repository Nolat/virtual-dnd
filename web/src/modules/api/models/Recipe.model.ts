import { Field, ID, ObjectType } from "type-graphql";
import { BaseEntity, Column, Entity, PrimaryGeneratedColumn } from "typeorm";

@ObjectType()
@Entity()
export class Recipe extends BaseEntity {
  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id: string;

  @Field()
  @Column()
  title: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  description?: string;
}
