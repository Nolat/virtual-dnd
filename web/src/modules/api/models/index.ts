export { default as User } from "./User.model";
export { default as Account } from "./Account.model";
export { default as Session } from "./Session.model";
export { default as Game } from "./Game.model";
export { default as GameUser } from "./GameUser.model";
