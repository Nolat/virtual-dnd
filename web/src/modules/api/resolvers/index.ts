export { GameResolver } from "./Game.resolver";
export { UserResolver } from "./User.resolver";
