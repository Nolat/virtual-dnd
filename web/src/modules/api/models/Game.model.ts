import { Field, ID, ObjectType } from "type-graphql";
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn
} from "typeorm";

import { GameUser, User } from "modules/api/models";

@Entity()
@ObjectType()
export default class Game extends BaseEntity {
  @PrimaryGeneratedColumn("uuid")
  @Field(() => ID)
  id: string;

  @Column()
  @Field()
  name: string;

  @ManyToOne(() => User)
  @Field()
  master: User;

  @OneToMany(() => GameUser, (gameUser) => gameUser.game)
  gameUsers: GameUser[];

  @Field(() => [User])
  users(): User[] {
    return this.gameUsers.map((gu) => gu.user);
  }

  @Column({ nullable: true })
  @Field({ nullable: true })
  password: string;

  @CreateDateColumn({ type: "timestamp" })
  @Field()
  createdAt: Date;

  @UpdateDateColumn({ type: "timestamp" })
  @Field()
  updatedAt: Date;
}
