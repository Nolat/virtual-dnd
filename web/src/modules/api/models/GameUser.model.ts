import { Field, ID, ObjectType } from "type-graphql";
import { BaseEntity, Column, Entity, ManyToOne, PrimaryGeneratedColumn } from "typeorm";

import { Game, User } from "modules/api/models";

@Entity()
@ObjectType()
export default class GameUser extends BaseEntity {
  @PrimaryGeneratedColumn("uuid")
  @Field(() => ID)
  id: string;

  @ManyToOne(() => User, (user) => user.games)
  @Field(() => User)
  user: User;

  @ManyToOne(() => Game, (game) => game.users)
  @Field()
  game: Game;

  @Column({ type: "timestamp", nullable: true })
  @Field({ nullable: true })
  lastSeenAt: Date;
}
