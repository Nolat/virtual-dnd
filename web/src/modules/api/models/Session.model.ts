import { randomBytes } from "crypto";

import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToOne,
  PrimaryGeneratedColumn,
  RelationId,
  UpdateDateColumn
} from "typeorm";

import { User } from "modules/api/models";

@Entity()
export default class Session extends BaseEntity {
  constructor() {
    super();
    this.sessionToken = randomBytes(32).toString("hex");
    this.accessToken = randomBytes(32).toString("hex");
  }

  @PrimaryGeneratedColumn("uuid")
  id: string;

  @OneToOne(() => User)
  @JoinColumn()
  user: User;

  @RelationId((session: Session) => session.user)
  userId: string;

  @Column({ type: "timestamp" })
  expiresOn: Date;

  @Column({ unique: true })
  sessionToken: string;

  @Column({ unique: true })
  accessToken: string;

  @CreateDateColumn({ type: "timestamp" })
  createdAt: Date;

  @UpdateDateColumn({ type: "timestamp" })
  updatedAt: Date;
}
