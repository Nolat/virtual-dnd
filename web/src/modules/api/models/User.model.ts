import { Field, ID, ObjectType } from "type-graphql";
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn
} from "typeorm";

import { Game, GameUser } from "modules/api/models";

@Entity()
@ObjectType()
export default class User extends BaseEntity {
  @PrimaryGeneratedColumn("uuid")
  @Field(() => ID)
  id: string;

  @Column()
  @Field()
  name: string;

  @Column({ unique: true, nullable: true })
  @Field()
  email: string;

  @Column()
  @Field()
  image: string;

  @OneToMany(() => GameUser, (gameUser) => gameUser.user)
  gameUsers: GameUser[];

  @Field(() => [Game], { nullable: true })
  games(): Game[] {
    return this.gameUsers.map((gu) => gu.game);
  }

  @CreateDateColumn({ type: "timestamp" })
  @Field()
  createdAt: Date;

  @UpdateDateColumn({ type: "timestamp" })
  @Field()
  updatedAt: Date;
}
