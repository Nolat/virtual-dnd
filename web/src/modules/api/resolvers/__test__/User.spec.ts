import { gql } from "apollo-server-micro";
import { createTestClient } from "apollo-server-testing";
import { getConnection } from "typeorm";

import { User } from "modules/api/models";
import { bootstrapTestServer } from "test/setup/bootstrapServer";

describe("User resolver", () => {
  const USER1_DATA = { email: "test1@test.com", name: "test1-user", image: "img:test1" };
  const USER2_DATA = { email: "test2@test.com", name: "test2-user", image: "img:test2" };

  const users: User[] = [];

  beforeAll(async () => {
    const connection = getConnection();
    const userRepository = connection.getRepository(User);

    const user1 = await userRepository.create(USER1_DATA);
    await user1.save();

    const user2 = await userRepository.create(USER2_DATA);
    await user2.save();

    users.push(user1, user2);
  });

  afterAll(async () => {
    const connection = getConnection();
    const userRepository = connection.getRepository(User);

    userRepository.remove(users);
  });

  it("should get a list of users", async () => {
    const server = await bootstrapTestServer({});

    const { query } = createTestClient(server);

    const res = await query({
      query: gql`
        query Users {
          Users {
            name
            email
            image
          }
        }
      `
    });

    expect(res.data.Users).toMatchSnapshot();
  });

  it("should get a specific user from id", async () => {
    const server = await bootstrapTestServer({});

    const { query } = createTestClient(server);

    const res = await query({
      query: gql`
        query User($id: String!) {
          User(id: $id) {
            name
            email
            image
          }
        }
      `,
      variables: { id: users[0].id }
    });

    expect(res.data.User).toMatchSnapshot();
  });

  it("should get a user when passed from context", async () => {
    const user = users[1];

    const server = await bootstrapTestServer({
      context: async () => {
        return { user };
      }
    });

    const { query } = createTestClient(server);

    const res = await query({
      query: gql`
        query me {
          me {
            name
            email
            image
          }
        }
      `
    });

    expect(res.data.me).toMatchSnapshot();
  });
});
