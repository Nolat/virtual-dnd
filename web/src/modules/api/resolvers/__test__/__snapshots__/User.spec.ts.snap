// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User resolver should get a list of users 1`] = `Array []`;

exports[`User resolver should get a specific user from id 1`] = `null`;

exports[`User resolver should get a user when passed from context 1`] = `
Object {
  "email": "test2@test.com",
  "image": "img:test2",
  "name": "test2-user",
}
`;
